File: 2018/02/advent201802.py
from collections import Counter

def part1(lines: list[str]) -> int:
  count2 = 0
  count3 = 0
  for line in lines:
    charCount = Counter(line)
    if 2 in charCount.values(): count2 += 1
    if 3 in charCount.values(): count3 += 1
  return count2 * count3

def part2(lines: list[str]) -> str:
  for i, line in enumerate(lines):
    for line2 in lines[i+1:]:
      diff = 0
      for j, char in enumerate(line2):
        if char != line[j]:
          diff += 1
          if diff > 1:
            break
      if diff == 1:
        res: list[str] = []
        for j, char in enumerate(line2):
          if char == line[j]:
            res.append(char)
        return ''.join(res)
  return ''

if __name__ == "__main__":
  file1 = open('input.txt','r')
  lines = [line.strip() for line in file1.readlines()]

  print(part1(lines))
  print(part2(lines))

File: 2018/2/advent201802.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Solver {
  private static List<String> parseInput() {
    String fileName = "input.txt";
    List<String> lines = new ArrayList<>();

    try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
      String line;
      while ((line = br.readLine()) != null) {
        lines.add(line);
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return lines;
  }

  private static HashMap<Character, Integer> makeMap(String line) {
    HashMap<Character, Integer> chars = new HashMap<>();
    for (Character c: line.toCharArray()) {
      if (chars.containsKey(c)) {
        chars.put(c, chars.get(c) + 1);
      } else {
        chars.put(c, 1);
      }
    }
    return chars;
  }

  private static int part1(List<String> lines) {
    int two = 0;
    int three = 0;
    for (String line: lines) {
      HashMap<Character, Integer> chars = makeMap(line);
      if (chars.containsValue(2)) {
        two++;
      }
      if (chars.containsValue(3)) {
        three++;
      }
    }
    return two * three;
  }

  private static String part2(List<String> lines) {
    for (int i = 0; i < lines.size(); i++) {
      char[] line1 = lines.get(i).toCharArray();
      for (int j = i + 1; j < lines.size(); j++) {
        char[] line2 = lines.get(j).toCharArray();
        int diff = 0;
        for (int k = 0; k < line1.length; k++) {
          if (line1[k] != line2[k]) {
            diff++;
            if (diff > 1) {
              break;
            }
          }
        }
        if (diff == 1) {
          StringBuilder res = new StringBuilder(line1.length - 1);
          for (int k = 0; k < line1.length; k++) {
            if (line1[k] == line2[k]) {
              res.append(line1[k]);
            }
          }
          return res.toString();
        }
      }
    }
    return "";
  }
  public static void main(String[] args) {
    List<String> lines = parseInput();
    System.out.println(part1(lines));
    System.out.println(part2(lines));

  }
}
